Support knx tunable white and color temperature lights

marvin-w opened the pull request:
## Description:

This PR intends to add support for tunable white and color temperature for knx lights. It does bump the version of the xknx library to `0.9.4`.

Closes #18266

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7979

cc @farmio

## Release notes

* Support for tunable white and color temperature for KNX lights.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Commit:
KNX: Bumped version to 0.9.4 and added support for tunable white and color temperature for lights.

Julius2342 commented:
This PR also incorporates all changes from  #18266 (the never ending story).

Commit:
Updated to the latest changes

marvin-w commented:
@farmio Can you please test the latest changes and give appropriate feedback? I don't own those kind of devices.

farmio commented:
works for me ;)
thank you for your help and the PR!

MartinHjelmare commented:
Is the plan that we merge this instead of #18266?

marvin-w commented:
@MartinHjelmare yes thats the plan

MartinHjelmare commented:
@amelchio you reviewed #18266. Can you have a look at this?

amelchio commented:
Okay, will do. Maybe not today, maybe not tomorrow, but soon.

amelchio reviewed (changes requested):
Looks good, I have just some small comments.

It also looks complicated but that seems inherent.

farmio commented:
@amelchio Thank you for your review! 
I second that Kelvin is the unit that is used in KNX and xknx (and most bulbs probably), but it seemed to me that Mireds is the only unit for color temperature in Home Assistant. So I tried to use Mireds in the HA-plugin as a main unit and Kelvin in the xknx device module.
Additionally I tried to minimize calls to `color_util.color_temperature_kelvin_to_mired()` and its counterpart - mostly for brevity and clarity.

farmio commented on homeassistant/components/light/knx.py:
```diff
@@ -111,26 +169,47 @@ def should_poll(self):
+        if self.device.supports_tunable_white:
+            relative_ct = self.device.current_tunable_white
+            return self.min_mireds + (((255 - relative_ct) / 255) *
+                                      (self.max_mireds - self.min_mireds)) \
```
As Mireds is the go-to unit for HA it seemed unnecessary to calculate this in a foreign unit. Especially since all Mireds values in this calculation are constants.

amelchio replied:
It matters. For your default ranges:

* 2700-6000 Kelvin, 50% is 4350 Kelvin
* 370-166 mireds, 50% is 268 mireds

However, 4350 Kelvin is 230 mireds.

amelchio replied:
The slider is indeed not intuitive because it is in mireds while we usually think in Kelvin.

farmio replied:
So shall we leave it as it is for lights that represent their color temperature in percent (outside of the HA world)? 
Or shall we go with the correct values but lesser intuitive sliders?

amelchio replied:
Do it right or you will get error reports that Home Assistant sets the wrong temperature (I have been there, several times).

Fixing the slider is an unrelated issue that could improve the UI for all lights.

farmio commented on homeassistant/components/light/knx.py:
```diff
@@ -154,19 +233,57 @@ def supported_features(self):
+            if kelvin > self._max_kelvin:
+                kelvin = self._max_kelvin
+            elif kelvin < self._min_kelvin:
+                kelvin = self._min_kelvin
```
I used your proposed approach in a previous version. It came to rounding errors when reaching _min_mireds. 
If I remember correctly: 
- User set min_kelvin to 2550 -> lowest possible value from HA-plugin to xknx after conversion and comparison in Mireds and conversion to Kelvin was 2551. 
- User set max_kelvin to 6200 -> highest value from HA-plugin to xknx after conversion and comparison in Mireds and conversion to Kelvin was 6211 which exceeds the user setting. 

I would like to leave the user defined temperature boundaries in Kelvin because it is the same Unit thats used in the KNX actuator parameterisation.

farmio replied:
We can't remove `self.min_mireds` and `self.max_mireds` propertys because they are needed to override the default properties of the `Light` class. We could remove the `self._min_mireds` attributes and put its calculation `color_util.color_temperature_kelvin_to_mired(self._max_kelvin)`
  in the `self.min_mireds` property.

farmio commented on homeassistant/components/light/knx.py:
```diff
@@ -154,19 +233,57 @@ def supported_features(self):
+        hs_color = kwargs.get(ATTR_HS_COLOR, self.hs_color)
+        mireds = kwargs.get(ATTR_COLOR_TEMP, self.color_temp)
+
+        # fall back to default values, if required
```
I'm sure @oliverblaha knows why.

farmio replied:
I think it is required when for RGB lights. Eg. when hs_color shall be changed but brightness is not yet known. Otherwise `color_util.color_hsv_to_RGB(*hs_color, brightness * 100 / 255)` in line 269 would fail. 
Can somebody confirm this - @Julius2342 maybe?

oliverblaha replied:
Sorry, missed that conversation before!
Yes, that's exactly when you need it. For converting to RGB, you always need hue, saturation and brightness; however, ha might send hue and saturation only. If that is the case, then the current brightness should be used. If that is not known, there's no other choice than taking some default value (or, well, just do nothing).

amelchio replied:
To be clear, I am not saying that the code is wrong. Just that "if required" carries no information and should be replaced with text actually saying when it is required.

farmio replied:
@oliverblaha thanks for your reply! So I guess we can move the fall back tests inside of 265
```python
elif self.device.supports_color and \
                (update_brightness or update_color):
            # change RGB color (includes brightness)
```
to make this clearer?

oliverblaha replied:
I would recommend against moving it in there. Technically it would (currently) be the best position, however, I believe it would be much harder to read - especially as the brightness might be set from three different sources (new value, existing value or default value).

That's why I chose to make sure that every variable is properly defined at the beginning of the method. There might be other scenarios later (e.g. when adding support for KNX devices that need HSV to set their color) which also need a default brightness if not set otherwise.

So I'd suggest to leave it where it is and change the comment as follows:
# to make sure that the following calculations always have valid values available, fall back to default values for any parameter that has neither been set to a new value nor has an known current value.

farmio replied:
Even if HSV devices were supported, HA would still address color changes the same way (through our update_color attribute). I would suggest to move it for now. We can always revert it if there is another usecase of this defaults.

farmio commented on homeassistant/components/light/knx.py:
```diff
@@ -111,26 +169,47 @@ def should_poll(self):
+        if self.device.supports_color_temperature:
+            kelvin = self.device.current_color_temperature
+            return color_util.color_temperature_kelvin_to_mired(kelvin) \
+                if kelvin is not None else DEFAULT_COLOR_TEMPERATURE
```
Tbh I just copied the brightness property here. I'll remove `DEFAULT_COLOR_TEMPERATURE` and test it. After all the KNX actuator should have a default on its own.

farmio committed:
return None when ct value is unknown

- return None instead of default value when ct is unknown
- remove DEFAULT_COLOR_TEMPERATURE

marvin-w committed:
Merge pull request #1 from farmio/feature/knx-added-support-for-tunable-white-and-color-temperature-for-lights

review change requests

homeassistant commented:
Hi @farmio, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#19699).

Once you do that we will be able to review and accept this pull request.

Thanks!

farmio committed:
use Kelvin as base for relative color temperature

use Kelvin as base for relative color temperature instead of Mireds

farmio committed:
moved fallback value tests for clarity

oliverblaha committed:
Update request from oliverblaha

Co-Authored-By: marvin-w <marvin@fam-wichmann.de>

farmio committed:
Merge branch 'feature/knx-added-support-for-tunable-white-and-color-temperature-for-lights' into feature/knx-added-support-for-tunable-white-and-color-temperature-for-lights

marvin-w committed:
Merge pull request #2 from farmio/feature/knx-added-support-for-tunable-white-and-color-temperature-for-lights

review change requests

amelchio commented:
This seems to be shaping up nicely but I lost track ... are we ready for a second review?

Please note the conflict, probably because the file has moved to `components/knx/light.py` in dev.

Commit:
Merge branch 'dev' into feature/knx-added-support-for-tunable-white-and-color-temperature-for-lights

amelchio commented on homeassistant/components/knx/light.py:
```diff
@@ -154,19 +236,59 @@ def supported_features(self):
+                (update_brightness or update_color):
+            # change RGB color (includes brightness)
+            # if brightness or hs_color was not yet set use the default value
+            # to claculate RGB from as a fallback
```
Typo: claculate

marvin-w replied:
fixed

amelchio reviewed (changes requested):
Looks good! Just one style issue I noticed about accessing `config` keys.

marvin-w commented on homeassistant/components/knx/light.py:
```diff
@@ -60,6 +90,20 @@ def async_add_entities_discovery(hass, discovery_info, async_add_entities):
+    group_address_tunable_white_state = None
+    group_address_color_temp = None
+    group_address_color_temp_state = None
+    if config.get(CONF_COLOR_TEMP_MODE) == ColorTempModes.absolute:
```
Isn't it optional according to the config:

```
vol.Optional(CONF_COLOR_TEMP_MODE, default=DEFAULT_COLOR_TEMP_MODE):
        cv.enum(ColorTempModes),
```

or am I getting this wrong

marvin-w replied:
Thanks for pointing that out, done!

farmio replied:
I think `group_address_switch` is required too.

Commit:
Address suggested changes

amelchio reviewed (approved):
Awesome 🎉

farmio commented on homeassistant/components/knx/light.py:
```diff
@@ -60,16 +90,36 @@ def async_add_entities_discovery(hass, discovery_info, async_add_entities):
         hass.data[DATA_KNX].xknx,
-        name=config.get(CONF_NAME),
+        name=config[CONF_NAME],
         group_address_switch=config.get(CONF_ADDRESS),
```
```suggestion
        group_address_switch=config[CONF_ADDRESS],
```

farmio replied:
`CONF_ADDRESS` is required too. See line 52
```suggestion
        group_address_switch=config[CONF_ADDRESS],
```

farmio commented on homeassistant/components/knx/light.py:
```diff
@@ -154,19 +236,59 @@ def supported_features(self):
+            # calculate relative_ct from Kelvin to fit typical KNX devices
+            kelvin = int(color_util.color_temperature_mired_to_kelvin(mireds))
+            relative_ct = int(255 * (kelvin - self._min_kelvin) /
+                                    (self._max_kelvin - self._min_kelvin))
```
pylint: Wrong continued indentation
https://travis-ci.org/home-assistant/home-assistant/jobs/490179399
```suggestion
                              (self._max_kelvin - self._min_kelvin))
```

farmio committed:
Update homeassistant/components/knx/light.py

Co-Authored-By: marvin-w <marvin@fam-wichmann.de>

farmio committed:
Update homeassistant/components/knx/light.py

Co-Authored-By: marvin-w <marvin@fam-wichmann.de>

farmio commented:
Hooray!! Thank you Marvin and Amelchio 😊

furiaceka87 commented:
After this update, we need to modify something for "normal" lights. Because in 0.88.0 I have KNX light.py issues (https://github.com/home-assistant/home-assistant/issues/21251)
